### PR TITLE
Gated feature flag to add a secondary outbound bind for IPv6-only clusters

### PR DESCRIPTION
**Pull request by pseudonator**
*Why this PR is required ?*

Currently in IPv6-only clusters such as EKS it introduces an egress-only IPv4 IP per pod that is fully capable for NAT'ing outbound traffic.

When running in these clusters applications should be able to route to IPv4 endpoints without having to enable full dual stack support.

However, as explained in https://github.com/istio/istio/issues/46719 when running Istio in the single stack IPv6 mode it only binds to,
```
tcp6   0      0 :::15001                :::*                   LISTEN      -      
```

To overcome this limitation this PR introduces a secondary bind IPv4 address to `15001`.

When this secondary bind is added this will create two listeners,
```
tcp        0      0 0.0.0.0:15001           0.0.0.0:*               LISTEN      25/envoy
tcp6       0      0 :::15001                :::*                    LISTEN      25/envoy
```

*Safety and Risk*

Currently this is feature gated with `ISTIO_ENABLE_IPV4_OUTBOUND_LISTENER_FOR_IPV6_CLUSTERS` flag so this secondary bind only gets enabled when this flag is set to `true`.

*Future Improvement*

I think we should also consider adding `ipv4_compat` (enabled in [Envoy socket address API](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-socketaddress)) for managing this compatibility in the future.

### CODE REVIEW

**Comment by howardjohn** [comment]
cc @jacob-delgado @zhlsunshine 

Is it *only* this one that needs two?

**Comment by pseudonator** [comment]
Correct,

Dual stack already manages these binds for multi interface environments

**Comment by zhlsunshine** [comment]
Hi @pseudonator, what I know is that: `::15001` should be okay for both IPv4 and IPv6 as a listener, so I am curious if there is other way to handle this?  You know it's very strange to add a ipv4 outbound listener to a IPv6-only cluster.

**Comment by pseudonator** [comment]
Thanks @zhlsunshine, I thought envoy by default does not handle this in IPv6 mode

Wasnt that the reason `ipv4_compat` was added ?
> ([bool](https://developers.google.com/protocol-buffers/docs/proto#scalar)) When binding to an IPv6 address above, this enables [IPv4 compatibility](https://tools.ietf.org/html/rfc3493#page-11). Binding to :: will allow both IPv4 and IPv6 connections, with peer IPv4 addresses mapped into IPv6 space as ::FFFF:<IPv4-address>.

**Comment by zhlsunshine** [comment]
> Thanks @zhlsunshine, I thought envoy by default does not handle this in IPv6 mode
> 
> Wasnt that the reason `ipv4_compat` was added ?
> 
> > ([bool](https://developers.google.com/protocol-buffers/docs/proto#scalar)) When binding to an IPv6 address above, this enables [IPv4 compatibility](https://tools.ietf.org/html/rfc3493#page-11). Binding to :: will allow both IPv4 and IPv6 connections, with peer IPv4 addresses mapped into IPv6 space as ::FFFF:.

Yeah, I remember that envoy has already supported `ipv4_compat` in the socket address, you may can refer to [this](https://github.com/istio/istio/pull/29076) for detail information.

**Comment by zhlsunshine** [comment]
> Thanks @zhlsunshine, I thought envoy by default does not handle this in IPv6 mode
> 
> Wasnt that the reason `ipv4_compat` was added ?
> 
> > ([bool](https://developers.google.com/protocol-buffers/docs/proto#scalar)) When binding to an IPv6 address above, this enables [IPv4 compatibility](https://tools.ietf.org/html/rfc3493#page-11). Binding to :: will allow both IPv4 and IPv6 connections, with peer IPv4 addresses mapped into IPv6 space as ::FFFF:.

Yeah, I remember that envoy has already supported `ipv4_compat` in the socket address,  so @pseudonator you may refer to [this](https://github.com/istio/istio/pull/29076) for detail information.

**Comment by pseudonator** [comment]
Thanks for referencing that PR. Introducing `ipv4_compat` is a big piece of work as i mentioned in the PR description.

And i totally get your concerns about adding a IPv4 bind in a IPv6-only environment but unfortunately thats what it comes down to in EKS environments. It still NATs traffic to both IPv4 and IPv6 in these environments. Hence why i have gated it so everyone dont need to opt-in to this.

I'm more than happy to address any further concerns you have.

**Comment by linsun** [comment]
> Hi @pseudonator, what I know is that: `::15001` should be okay for both IPv4 and IPv6 as a listener, so I am curious if there is other way to handle this? You know it's very strange to add a ipv4 outbound listener to a IPv6-only cluster.

It is actually not strange if you know other clusters are IPv4 around the few IPv6-only clusters.  @pseudonator do you have users want to do this, could be helpful if you can share?

@zhlsunshine do you see any issue of dark launch this to get users feedback first?

**Comment by linsun** [comment]
@jacob-delgado @ilrudie thoughts?

**Comment by ilrudie** [comment]
This looks reasonable. Using `additionalAddresses` keeps the scope pretty well limited

**Comment by pseudonator** [comment]
Any further feedback on this PR ?

**Comment by ilrudie** [comment]
@zhlsunshine, your input on the idea of a dark launch and the concerns that would prevent moving forward with this would be greatly appreciated.

**Comment by zhlsunshine** [comment]
Hi @linsun & @ilrudie, I have no concern about the performance with 2 listeners, and @pseudonator I also have no more concern for this PR. There is no doubt that this PR can handle the issue in EKS environment.  And I suggest that we can add the issue or PR link in the place where the code is changed.

**Comment by pseudonator** [comment]
@zhlsunshine got it thank you. I will add the issue link as a comment

**Comment by jacob-delgado** [comment]
This LGTM, but does this also need a change for `istio-cni`?

**Comment by pseudonator** [comment]
looks like tests are good. over to you @zhlsunshine @jacob-delgado

**Comment by pseudonator** [comment]
@zhlsunshine @jacob-delgado any further feedback or update on this ? dont want this PR to fall behind

**Comment by mmerickel** [comment]
> Dual stack already manages these binds for multi interface environments

I have not observed this to be the case when setting ISTIO_DUAL_STACK=true - am I missing something? Is it unreleased? To me it makes sense to consider this scenario of IPv6-only EKS deployed in a dual stack environment where pods want to still perform IPv4 traffic as "dual stack".... but again the ISTIO_DUAL_STACK=true does not allow IPv4 traffic to exit the app containers in this EKS scenario.

Ultimately I'm asking why this flag is needed - as a user my first impression was that I expected ISTIO_DUAL_STACK=true to apply to this scenario and do the right thing.

**Comment by pseudonator** [comment]
@mmerickel I can chime in a bit here, 

- In the case of IPv6-only EKS its not a true dual stack in the sense the IP family is single stack IPv6 only. And if you need IPv4 capability for SE then IPv6 binding alone does not help. This flag is a means to gate this behaviour such that it does not introduce unnecessary binds in other cloud providers. as far as i know its only in EKS this IPv4 NAT scenario exists for outbound traffic.
- When it comes to dual stack i don't think its quite GA yet. So your mileage will vary if you use `ISTIO_DUAL_STACK` flag. 
WG network folks can add more colour to this.

**Comment by mmerickel** [comment]
> I would still want to define ServiceEntry objects for services that only support IPv4 in this EKS config.

@pseudonator So will I be able to do this with this PR, with dual stack disabled? And view mesh traffic to those services?

**Comment by pseudonator** [comment]
That's the intention 🙂

**Comment by zhlsunshine** [comment]
Hi @pseudonator LGTM, thanks!
BTW, want to know that: is it okay if there is a service entry which is defined with only ipv6 address in EKS environment?

**Review by jacob-delgado — APPROVED** [review]

**Comment by pseudonator** [comment]
Thanks @jacob-delgado 

@zhlsunshine I assume you mean something like this right ?
```
spec:
  hosts:
    - example.com
  location: MESH_EXTERNAL
  ports:
    - number: 443
      name: https
      protocol: HTTPS
  resolution: STATIC
  endpoints:
    - address: "2606:2800:220:1:248:1893:25c8:1946"
```

if so yes it does

Heres an output of a route to both `93.184.216.34` and `2606:2800:220:1:248:1893:25c8:1946` addresses
Screenshot 2023-10-02 at 21 50 52

**Review by zhlsunshine — APPROVED** [review]
Thanks @pseudonator, LGTM!

**Comment by pseudonator** [comment]
Thanks @zhlsunshine

**Comment by ilrudie** [comment]
@howardjohn, Jacob and Steve approved. Any chance you could take another look at this?

File: pilot/pkg/features/pilot.go [hunk 1]
```diff
@@ -680,6 +680,11 @@ var (
 	EnableSelectorBasedK8sGatewayPolicy = env.Register("ENABLE_SELECTOR_BASED_K8S_GATEWAY_POLICY", true,
 		"If disabled, Gateway API gateways will ignore workloadSelector policies, only"+
 			"applying policies that select the gateway with a targetRef.").Get()
+
+	// Useful for IPv6-only EKS clusters. See https://aws.github.io/aws-eks-best-practices/networking/ipv6/ why it assigns an additional IPv4 NAT address.
+	// Also see https://github.com/istio/istio/issues/46719 why this flag is required
+	EnableAdditionalIpv4OutboundListenerForIpv6Only = env.RegisterBoolVar("ISTIO_ENABLE_IPV4_OUTBOUND_LISTENER_FOR_IPV6_CLUSTERS", false,
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/features/pilot.go:686]: Having EKS IPv6 users need to set the flag indefinitely isn't really an acceptable UX long term is my concern. Fine to have the flag for short term though

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/features/pilot.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pseudonator) [reply pilot/pkg/features/pilot.go:686]: @howardjohn all good points you raise. Will tackle this in a separate PR. 

Theres also a question of how Ambient will deal with this if and when IPv6 lands

**Comment by linsun** [comment]
Any objection of cherry pick this fix to release 1.19? @dims from AWS also pinged me about how this fix help them as well.

cc @pseudonator

**Comment by linsun** [comment]
Hi @pseudonator can you check release-1.20 branch has this fix? Just realized branching was cut also around 2 weeks ago.

**Comment by mmerickel** [comment]
I'm running into issues where ServiceEntry telemetry is not tagged properly for ipv4 egress traffic using this flag - thought I'd link it here to see if anyone else has run into this: https://github.com/istio/istio/issues/48829

**Pull request merged**